agent/lib/kontena/workers/service_pod_worker.rb

```diff
@@ -128,17 +130,53 @@ def apply
           warn "failed to sync #{service_pod.name} at #{service_pod.deploy_rev}: #{error}"
           warn error
           sync_state_to_master(@container, error)
+          return
         else
           @container_state_changed = false
-          sync_state_to_master(@container)
+        end
 
+        if service_pod.terminated?
           # Only terminate this actor after we have succesfully ensure_terminated the Docker container
           # Otherwise, stick around... the manager will notice we're still there and re-signal to destroy
-          self.terminate if service_pod.terminated?
+          self.terminate
+        elsif service_pod.running? && service_pod.wait_for_port
+          # delay sync_state_to_master until started
+          # XXX: apply() gets called twice for each deploy_rev, and this launches two wait_for_port tasks...
+          async.wait_for_port(service_pod, @container)
```

Comment:
I don't like that `wait_for_port` is such a special case here.. could it return state (container) that is reported normally back to master (not inside wait_for_port)?

Reply:
Or maybe this could be a separate class/actor that checks pod readiness.

Reply:
WIP on a health check actor managed by the service pod worker that tracks the service container ready/healthy state, where the server deploy waits for the service instance to go `initialized` -> `starting` -> `running`...

Reply:
> I don't like that wait_for_port is such a special case here.. could it return state (container) that is reported normally back to master (not inside wait_for_port)?

It has to be an `async` call if it's in the `exclusive` block, and I'm hesitant to break apart the `exclusive { ... sync_state_to_master }` block because the way the `@service_pod` works is rather unclear. It has to avoid calling `sync_state_to_master` if the `@service_pod` has changed, because that would report the status of the old service rev with the newer deploy rev.

Maybe it would be better to use the same non-exclusive `if @service_pod.deploy_rev == service_pod.deploy_rev` check in all cases, though, both with and without `wait_for_port`.

Reply:
Changed to move the `sync_state_to_master` out of the `exclusive` block.